include/oneapi/dpl/pstl/algorithm_ranges_impl.h

```diff
@@ -349,6 +350,325 @@ __pattern_search_n(_Tag __tag, _ExecutionPolicy&& __exec, _R&& __r, std::ranges:
+//---------------------------------------------------------------------------------------------------------------------
+
+template <typename _Tag, typename _ExecutionPolicy, typename _InRange, typename _OutRange>
+void
+__pattern_copy_impl(_Tag __tag, _ExecutionPolicy&& __exec, _InRange&& __in_r, _OutRange&& __out_r)
+{
+    static_assert(__is_parallel_tag_v<_Tag> || typename _Tag::__is_vector{});
+
+    assert(std::ranges::size(__in_r) == std::ranges::size(__out_r));
+
+    oneapi::dpl::__internal::__pattern_walk2_brick(__tag, std::forward<_ExecutionPolicy>(__exec),
+        std::ranges::begin(__in_r), std::ranges::begin(__in_r) + std::ranges::size(__in_r), std::ranges::begin(__out_r),
+        oneapi::dpl::__internal::__brick_copy<decltype(__tag), _ExecutionPolicy>{});
+}
+
+template <typename _IsVector, typename _ExecutionPolicy, typename _InRange, typename _OutRange>
+void
+__pattern_copy(__parallel_tag<_IsVector> __tag, _ExecutionPolicy&& __exec, _InRange&& __in_r, _OutRange&& __out_r)
+{
+    __pattern_copy_impl(__tag, std::forward<_ExecutionPolicy>(__exec), std::forward<_InRange>(__in_r),
+        std::forward<_OutRange>(__out_r));
+}
+
+template<typename _Tag, typename _ExecutionPolicy, typename _InRange, typename _OutRange>
+void
+__pattern_copy(_Tag __tag, _ExecutionPolicy&& __exec, _InRange&& __in_r, _OutRange&& __out_r)
+{
+    if constexpr(typename _Tag::__is_vector{})
+        __pattern_copy_impl(__tag, std::forward<_ExecutionPolicy>(__exec), std::forward<_InRange>(__in_r),
+            std::forward<_OutRange>(__out_r));
+    else
+        std::ranges::copy(std::forward<_InRange>(__in_r), std::ranges::begin(__out_r));
+}
+//---------------------------------------------------------------------------------------------------------------------
+// pattern_copy_if
+//---------------------------------------------------------------------------------------------------------------------
+
+template <typename _Tag, typename _ExecutionPolicy, typename _InRange, typename _OutRange, typename _Pred,
+          typename _Proj>
+auto
+__pattern_copy_if_impl(_Tag __tag, _ExecutionPolicy&& __exec, _InRange&& __in_r, _OutRange&& __out_r, _Pred __pred, _Proj __proj)
+{
+    static_assert(__is_parallel_tag_v<_Tag> || typename _Tag::__is_vector{});
+
+    auto __pred_1 = [__pred, __proj](auto&& __val) { return std::invoke(__pred, std::invoke(__proj,
+        std::forward<decltype(__val)>(__val)));};
+
+    auto __res_idx = oneapi::dpl::__internal::__pattern_copy_if(__tag, std::forward<_ExecutionPolicy>(__exec),
+        std::ranges::begin(__in_r), std::ranges::begin(__in_r) + std::ranges::size(__in_r),
+        std::ranges::begin(__out_r), __pred_1) - std::ranges::begin(__out_r);
+
+    using __return_t = std::ranges::copy_if_result<std::ranges::borrowed_iterator_t<_InRange>,
+        std::ranges::borrowed_iterator_t<_OutRange>>;
+
+    return __return_t{std::ranges::begin(__in_r) + std::ranges::size(__in_r), __out_r.begin() + __res_idx};
```

Comment:
Should we use `std::ranges::begin(__out_r)` here instead of the member function?